### ISSUE TITLE
Thread safety

### ISSUE THREAD

**Issue by jnicklas** [teamcapybara/capybara#1248]
Capybara is not thread safe in that we rely on global configuration which is liable to change at any point. Unfortunately, fixing this is probably going to break APIs in a few subtle ways. Changing a global configuration parameter is not going to affect an already existing session, since we'd need to copy configuration to sessions to prevent concurrent modification, as well as providing an API to change options per session. I think this is consistent with how _most_ people use Capybara, but it's going to break for some people, so I want to do this for 3.0.

**Comment by timlentse** [comment]
Is this ok?
```ruby
5.times do |id|
    workers << Thread.new do
    session[id].visit(url)
   end
end
```
Here `session` is an array of session instance, I have tried, but it turned out to be dead lock. I don't know why?

**Comment by oystersauce8** [comment]
is there a branch or a patch that makes this possible?  When is capybara 3 coming out.

**Comment by twalpole** [comment]
There is no branch or patch currently and not current timeline for Capybara 3

**Comment by twalpole** [comment]
The largest part of this (the per session configuration) is being worked on in the `session-config` branch/PR #1808  if anyone has any feedback/suggestions

**Comment by twalpole** [comment]
Closed via 0270b7a4cf960055a87f93db1322b53204aec350
